pkg/pusher/pusher.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pusher
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+type Service struct {
+	storer            storage.Storer
+	peerSuggester     topology.ClosestPeerer
+	pushSyncer        pushsync.PushSyncer
+	logger            logging.Logger
+	metrics           metrics
+	quit              chan struct{}
+	chunksWorkerQuitC chan struct{}
+}
+
+type Options struct {
+	Storer        storage.Storer
+	PeerSuggester topology.ClosestPeerer
+	PushSyncer    pushsync.PushSyncer
+	Logger        logging.Logger
+}
+
+var retryInterval = 10 * time.Second // time interval between retries
+
+func New(o Options) *Service {
+	service := &Service{
+		storer:            o.Storer,
+		peerSuggester:     o.PeerSuggester,
+		pushSyncer:        o.PushSyncer,
+		logger:            o.Logger,
+		metrics:           newMetrics(),
+		quit:              make(chan struct{}),
+		chunksWorkerQuitC: make(chan struct{}),
+	}
+	go service.chunksWorker()
+	return service
+}
+
+// chunksWorker is a loop that keeps looking for chunks that are locally uploaded ( by monitoring pushIndex )
+// and pushes them to the closest peer and get a receipt.
+func (s *Service) chunksWorker() {
+	var chunks <-chan swarm.Chunk
+	var unsubscribe func()
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	defer close(s.chunksWorkerQuitC)
+	chunksInBatch := -1
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-s.quit
+		cancel()
+	}()
+	for {
+		select {
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				var dur time.Duration
+				if chunksInBatch == 0 {
+					dur = 500 * time.Millisecond
+				}
+				timer.Reset(dur)
+				break
+			}
+
+			chunksInBatch++
+			s.metrics.TotalChunksToBeSentCounter.Inc()
+			peer, err := s.peerSuggester.ClosestPeer(ch.Address())
+			if err != nil {
+				if errors.Is(err, topology.ErrWantSelf) {
+					// set chunk status to synced
+					s.setChunkAsSynced(ctx, ch.Address())
+					continue
+				}
+			}
+
+			// Later when we process receipt, get the receipt and process it
+			// for now ignoring the receipt and checking only for error
+			_, err = s.pushSyncer.ChunkPusher(ctx, peer, ch)
```

Comment:
Thank you @acud for spotting this improvement, that I actually missed while writing the the task description.

I agree that peerSuggester is not needed in pusher and that `pushSyncer.ChunkPusher` method can accept only a chunk and deal with first peer, just as it deals with the forwarding. This would remove pusher dependency peerSuggester, which is used only here. I am also suggesting to change the name of the method `ChunkPusher` from a noun to something with a ver, like, `PushChunk`, or `PushChunkToClosestPeer`, maybe there could be something prettier.

Reply:
Thanks @janos. I am removing the topology from pusher per our discussion.